lib/Language/ASL/Formulas/Normalize.hs

```diff
@@ -975,3 +977,93 @@ extractInts e = do
         expr'' <- extractBitV' expr'
         withSym $ \sym -> WI.sbvToInteger sym expr''
       _ -> return expr'
+
+data CondCache t tp = CondCache (IO.IORef (Map (WB.Expr t WI.BaseBoolType) (WB.Expr t tp)))
+
+-- | Path-sensitive cached evaluation
+condCacheEval :: IO.MonadIO m => WB.IdxCache t (CondCache t) -> WB.Expr t WI.BaseBoolType -> WB.Expr t tp -> m (WB.Expr t tp) -> m (WB.Expr t tp)
```

Comment:
Minor style nit pick: this line is pretty long, so it would be a bit easier to read if it were split up into multiple lines:

```suggestion
condCacheEval ::
  IO.MonadIO m =>
  WB.IdxCache t (CondCache t) ->
  WB.Expr t WI.BaseBoolType ->
  WB.Expr t tp ->
  m (WB.Expr t tp) ->
  m (WB.Expr t tp)
```

It might also be helpful to add Haddocks to describe what each argument does.